File: macos/LookoutMarine/ChartLinkFetch.swift
//  ChartLinkFetch.swift
//
//  The shell's whole part in charts by link: fetch the bytes at a url.
//
//  Probing the link, inlining TileJSON sources, generating a wrapper style for
//  bare tiles, fetching sprite packs, building the credit line, templating tile
//  urls and keeping the list are all lookout's. See lookout_set_http_provider
//  in lookout.h.
//
//  THREADING. `get` is called by lookout with its lock held, so it does the
//  least possible: start the task, return. `lookout_http_respond` is the one
//  call that is safe from there and from the URLSession completion, because it
//  takes no lock of lookout's — it enqueues, and the next frame adopts.

import Foundation

final class ChartLinkFetch: @unchecked Sendable {
    /// Say who is asking, on every chart-link request. Public tile hosts serve
    /// "access blocked" placeholder tiles to anonymous or platform-default
    /// agents — openstreetmap.org's tile usage policy
    /// (osm.wiki/Blocked_tiles) wants a unique, identifiable User-Agent with a
    /// way to reach the developer, and the Referer names the app's home for
    /// hosts that key on it.
    static let userAgent =
        "LookoutMarine/1.0 (macOS; org.beetlebug.lookout; contact jeremy.collins@beetlebug.org)"
    static let referer = "https://beetlebug.org/"

    private let lock = NSLock()
    /// nil once the chart is closing: a fetch still in flight must not answer
    /// into a handle that is going away.
    private var handle: OpaquePointer?
    /// Tasks by request id, so a cancel can reach the transfer.
    private var inFlight: [UInt64: URLSessionTask] = [:]
    private let session: URLSession

    init() {
        let cfg = URLSessionConfiguration.default
        // Tiles are the one thing this app fetches in bulk. A shared memory
        // cache spares the pan back over water already crossed, and the disk
        // cache survives a relaunch on the same chart. URL-keyed, so it covers
        // styles and sprite sheets on the same terms.
        cfg.requestCachePolicy = .useProtocolCachePolicy
        cfg.urlCache = URLCache(memoryCapacity: 16 << 20, diskCapacity: 256 << 20)
        // A stalled fetch must not hold a slot forever: the chart is drawn from
        // whatever HAS landed, so a slow tile costs only itself.
        cfg.timeoutIntervalForRequest = 20
        // URLSession pools per host. Connection concurrency is its business:
        // nothing here reasons about which source a url belongs to, so no
        // source can hold a lane another source's tiles are waiting on.
        cfg.httpMaximumConnectionsPerHost = 8
        cfg.httpAdditionalHeaders = [
            "User-Agent": ChartLinkFetch.userAgent,
            "Referer": ChartLinkFetch.referer,
        ]
        session = URLSession(configuration: cfg)
    }

    /// Attach to a chart handle and start answering. Call once per handle.
    func attach(to h: OpaquePointer) {
        lock.lock()
        handle = h
        lock.unlock()
        lookout_set_http_provider(h, chartLinkGet, chartLinkCancel,
                                  Unmanaged.passUnretained(self).toOpaque())
    }

    /// Stop answering, before the handle closes. Idempotent.
    ///
    /// The in-flight ids are answered "failed" FIRST: lookout holds an
    /// outstanding-request slot for every id it has neither been answered nor
    /// cancelled on, and clearing the provider is what releases the rest.
    func detach() {
        lock.lock()
        let h = handle
        let tasks = inFlight
        inFlight = [:]
        if let h {
            for (id, _) in tasks { lookout_http_respond(h, id, nil, 0, 0) }
        }
        handle = nil
        lock.unlock()
        if let h { lookout_set_http_provider(h, nil, nil, nil) }
        for (_, t) in tasks { t.cancel() }
    }

    /// One url lookout wants. Called on its render thread with its lock held —
    /// do not block.
    fileprivate func fetch(id: UInt64, url raw: String, allowFile: Bool) {
        guard let url = URL(string: raw) else {
            answer(id, nil, 0)
            return
        }
        // The file:// boundary. lookout says when a url may be read off disk
        // (see lookout_http_get): the link the mariner typed, and what a
        // document ALREADY read from disk names inside that link's directory.
        // A style that arrived over the network never gets it, so it cannot
        // make this read arbitrary local files as its "TileJSON".
        if url.isFileURL || raw.hasPrefix("/") {
            guard allowFile else {
                answer(id, nil, 0)
                return
            }
            readFile(id: id, url: url.isFileURL ? url : URL(fileURLWithPath: raw))
            return
        }
        let task = session.dataTask(with: URLRequest(url: url)) { [weak self] data, resp, err in
            guard let self else { return }
            self.done(id)
            let code = (resp as? HTTPURLResponse)?.statusCode ?? 0
            if err != nil {
                self.answer(id, nil, 0)
            } else {
                self.answer(id, data, Int32(code))
            }
        }
        lock.lock()
        if handle != nil { inFlight[id] = task }
        let live = handle != nil
        lock.unlock()
        if live { task.resume() } else { answer(id, nil, 0) }
    }

    /// A style file the mariner has aboard. Read off the main thread: a file
    /// read on a sandboxed volume can take long enough to be felt as a freeze.
    private func readFile(id: UInt64, url: URL) {
        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
            guard let self else { return }
            // The sandbox hands access over with the pick and takes it back
            // when the scope closes. Harmless where no scope is held.
            let scoped = url.startAccessingSecurityScopedResource()
            defer { if scoped { url.stopAccessingSecurityScopedResource() } }
            guard let data = try? Data(contentsOf: url) else {
                self.answer(id, nil, 0)
                return
            }
            self.answer(id, data, 200)
        }
    }

    fileprivate func abort(id: UInt64) {
        lock.lock()
        let task = inFlight.removeValue(forKey: id)
        lock.unlock()
        task?.cancel()
    }

    private func done(_ id: UInt64) {
        lock.lock()
        inFlight.removeValue(forKey: id)
        lock.unlock()
    }

    private func answer(_ id: UInt64, _ bytes: Data?, _ status: Int32) {
        // Under the lock, so a handle closing mid-answer cannot be answered
        // into. lookout_http_respond takes no lock of its own, so nothing can
        // deadlock behind this.
        lock.lock()
        defer { lock.unlock() }
        guard let h = handle else { return }
        if let bytes, !bytes.isEmpty {
            bytes.withUnsafeBytes { raw in
                lookout_http_respond(h, id, raw.baseAddress, raw.count, status)
            }
        } else {
            lookout_http_respond(h, id, nil, 0, status)
        }
    }
}

/// The C entry points. The url is lookout's memory and is only valid for the
/// call, so it is copied out at once.
private let chartLinkGet: lookout_http_get = { user, id, url, allowFile in
    guard let user, let url else { return }
    let f = Unmanaged<ChartLinkFetch>.fromOpaque(user).takeUnretainedValue()
    f.fetch(id: id, url: String(cString: url), allowFile: allowFile != 0)
}

private let chartLinkCancel: lookout_http_cancel = { user, id in
    guard let user else { return }
    Unmanaged<ChartLinkFetch>.fromOpaque(user).takeUnretainedValue().abort(id: id)
}

File: android/app/src/main/java/org/beetlebug/lookout/charts/ChartLinkFetch.kt
package org.beetlebug.lookout.charts

import org.beetlebug.lookout.Lookout

import android.util.Log
import java.io.File
import java.net.HttpURLConnection
import java.net.URL
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

/**
 * The shell's whole part in charts by link: fetch the bytes at a url.
 *
 * Probing the link, inlining TileJSON sources, generating a wrapper style for
 * bare tiles, fetching sprite packs, building the credit line, templating tile
 * urls and keeping the list are all lookout's. See lookout_set_http_provider
 * in lookout.h.
 *
 * THE JNI LAYER STAYS DOWNCALL-ONLY. The core raises its asks on the render
 * thread with the api lock held, so they park in a ring here instead of
 * upcalling into the JVM: one thread drains the ring, a small pool fetches, and
 * every ask is ANSWERED, because an id that is neither answered nor cancelled
 * holds one of the core's outstanding-request slots.
 */
class ChartLinkFetch {
    @Volatile private var live = false

    /**
     * Serializes httpRespond against stop(). A fetch that has passed the live
     * check must finish its JNI call before stop() returns, so the engine
     * handle cannot be freed while an answer is still in flight. The network
     * read happens before the lock is taken, so stop() waits milliseconds at
     * most.
     */
    private val respondLock = Any()
    private var engine: Lookout? = null
    private var pool: ExecutorService? = null
    private var poller: Thread? = null

    /**
     * Wakes the frame loop. An answer is adopted at the top of a frame, and the
     * loop stands down when nothing is moving, so a resolve landing with no
     * gesture behind it needs someone to ask for the next frame.
     */
    @Volatile private var wake: (() -> Unit)? = null

    /** In-flight connections by request id, so a cancel can reach one. */
    private val inFlight = ConcurrentHashMap<Long, HttpURLConnection>()

    @Synchronized
    fun start(l: Lookout, wake: () -> Unit) {
        if (live && engine === l) return
        stop()
        engine = l
        this.wake = wake
        live = true
        l.httpProvider(true)
        // A level change asks tens of tiles across several hosts; six workers
        // keep the tail short. No per-host throttle: a couple of slow or dead
        // requests to one host must not hold a lane and starve the rest.
        pool = Executors.newFixedThreadPool(6)
        poller = Thread({
            val ids = LongArray(16)
            val allow = IntArray(16)
            val urls = arrayOfNulls<String>(16)
            val cancelled = LongArray(64)
            while (live) {
                val c = l.httpCancelPoll(cancelled)
                for (i in 0 until c) inFlight.remove(cancelled[i])?.disconnect()
                val n = l.httpPoll(ids, allow, urls)
                if (n == 0 && c == 0) {
                    try { Thread.sleep(POLL_MS) } catch (_: InterruptedException) { break }
                    continue
                }
                for (i in 0 until n) {
                    val id = ids[i]
                    val url = urls[i] ?: ""
                    val allowFile = allow[i] != 0
                    pool?.execute { fetch(l, id, url, allowFile) }
                }
            }
        }, "lookout-chart-links").also { it.start() }
    }

    @Synchronized
    fun stop() {
        val l = engine
        live = false
        poller?.let { p -> p.interrupt(); try { p.join(1_000) } catch (_: InterruptedException) {} }
        poller = null
        pool?.shutdownNow()
        pool = null
        for ((_, conn) in inFlight) conn.disconnect()
        inFlight.clear()
        // After this point no new respond can start, because live is read under
        // the lock, and none is mid-call, because taking the lock waits for the
        // last one to finish.
        synchronized(respondLock) {}
        l?.httpProvider(false)
        engine = null
        wake = null
    }

    /** Every answer funnels through here; see respondLock. */
    private fun respond(l: Lookout, id: Long, bytes: ByteArray?, status: Int) {
        synchronized(respondLock) {
            if (!live) return
            l.httpRespond(id, bytes, status)
        }
        wake?.invoke()
    }

    /** One piece of an answer, same lock. Returns false once the fetcher has
     *  stood down, and the read stops there. */
    private fun respondPiece(
        l: Lookout,
        id: Long,
        buf: ByteArray,
        len: Int,
        status: Int,
        done: Boolean,
    ): Boolean {
        synchronized(respondLock) {
            if (!live) return false
            l.httpRespondChunk(id, buf, len, status, done)
        }
        wake?.invoke()
        return true
    }

    private fun fetch(l: Lookout, id: Long, url: String, allowFile: Boolean) {
        // The file:// boundary. lookout says when a url may be read off disk
        // (see lookout_http_get): the link the mariner typed, and what a
        // document already read from disk names inside that link's directory. A
        // style that arrived over the network never gets it, so it cannot make
        // this read arbitrary local files as its "TileJSON".
        val path = localPath(url)
        if (path != null) {
            if (!allowFile) {
                respond(l, id, null, 0)
                return
            }
            val bytes = try { File(path).readBytes() } catch (e: Exception) {
                Log.w(TAG, "chart link $url: $e")
                respond(l, id, null, 0)
                return
            }
            respond(l, id, bytes, 200)
            return
        }
        var conn: HttpURLConnection? = null
        try {
            conn = URL(url).openConnection() as HttpURLConnection
            conn.setRequestProperty("User-Agent", USER_AGENT)
            conn.setRequestProperty("Referer", REFERER)
            // Short timeouts: the chart is drawn from whatever HAS landed, and
            // a style that asks a base map past the zoom it actually serves
            // must fail fast rather than hold a worker.
            conn.connectTimeout = 8_000
            conn.readTimeout = 8_000
            inFlight[id] = conn
            val code = conn.responseCode
            if (code !in 200..299) {
                respond(l, id, null, code)
                return
            }
            // A piece at a time, into one buffer. A NOAA district downloads as
            // a zip of a couple of hundred megabytes, and read into a ByteArray
            // that is more than a phone's heap allows.
            val buf = ByteArray(READ_CHUNK)
            conn.inputStream.use { input ->
                while (true) {
                    val n = input.read(buf)
                    if (n < 0) break
                    if (n == 0) continue
                    if (!respondPiece(l, id, buf, n, code, false)) return
                }
            }
            respondPiece(l, id, buf, 0, code, true)
        } catch (e: Exception) {
            respond(l, id, null, 0)
        } finally {
            inFlight.remove(id)
            conn?.disconnect()
        }
    }

    /** The path a local url names, or null when it names a host. */
    private fun localPath(url: String): String? = when {
        url.startsWith("file://") -> url.removePrefix("file://")
        url.startsWith("/") -> url
        else -> null
    }

    companion object {
        private const val TAG = "lookout"

        /** How long the poller waits on an empty ring. Fast enough that a
         *  resolve does not feel stepped through, cheap enough to leave
         *  running while a link is up. */
        private const val POLL_MS = 16L

        /** How much of a body goes in one piece. Big enough that a 200 MB
         *  district is under a thousand handoffs, small enough that four of
         *  them in flight are a megabyte of buffers. */
        private const val READ_CHUNK = 256 * 1024

        /**
         * Say who is asking, on every chart-link request. Public tile hosts
         * serve "access blocked" placeholder tiles to anonymous or
         * platform-default agents — openstreetmap.org's tile usage policy
         * (osm.wiki/Blocked_tiles) wants a unique, identifiable User-Agent
         * with a way to reach the developer, and the Referer names the app's
         * home for hosts that key on it.
         */
        const val USER_AGENT =
            "LookoutMarine/1.0 (Android; org.beetlebug.lookout; contact jeremy.collins@beetlebug.org)"
        const val REFERER = "https://beetlebug.org/"
    }
}
